Web Inspector: Parse attributes added via editing the tag name

rcaliman-apple opened the pull request:
#### 0b084355708b8e18ed3953423fc6e7fc018cca90
<pre>
Web Inspector: Parse attributes added via editing the tag name
<a href="https://bugs.webkit.org/show_bug.cgi?id=276533">https://bugs.webkit.org/show_bug.cgi?id=276533</a>
<a href="https://rdar.apple.com/problem/131607290">rdar://problem/131607290</a>

Reviewed by Devin Rousso.

Allow the user to input attributes when editing a node&apos;s tag name in the DOM node tree.

Take the first space-delimited substring to use as the tag name and use the rest as a string of attributes.
If the new tag name is invalid, the entire transaction will reject and revert to the original node.

* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype._startEditingTagName.keyupListener):
(WI.DOMTreeElement.prototype._tagNameEditingCommitted):
(WI.DOMTreeElement.prototype._tagNameEditingCommitted.changeTagNameCallback): Deleted.

Canonical link: <a href="https://commits.webkit.org/281507@main">https://commits.webkit.org/281507@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02cba9c1a9dd19ad1f6137cef7a7eba24bd0b691

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59631 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38976 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12156 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63547 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10155 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46629 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10309 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48380 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7112 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61661 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36379 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51621 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29218 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33082 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8869 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9078 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55032 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9147 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65278 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3559 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9036 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55723 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3570 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51610 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55853 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13328 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2955 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34790 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35873 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36959 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35618 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -1057,7 +1057,7 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
         {
             if (closingTagElement)
-                closingTagElement.textContent = "</" + tagNameElement.textContent + ">";
+                closingTagElement.textContent = "</" + tagNameElement.textContent.trim().split(" ")[0] + ">";
```
rather than split the entire string, i think it'd be more efficient to do something like this
```js
if (closingTagElement) {
    let textContent = tagNameElement.textContent.trimStart();
    let whitespaceIndex = textContent.search(/\s/);
    if (whitespaceIndex === -1)
        textContent = textContent.slice(0, whitespaceIndex);
    closingTagElement.textContent = "</" + textContent + ">";
```

rcaliman-apple replied:
Neat! I forgot about `trimStart()`. But in this case `trim()` is still suitable to remove any stray spaces before and after a single new string.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -1271,18 +1271,37 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
-        var wasExpanded = this.expanded;
+        let treeOutline = this.treeOutline;
+        let wasExpanded = this.expanded;
+        let [newTagName, ...attributes] = newText.trim().split(" ");
```
what if there's multiple spaces in a row?

rcaliman-apple replied:
That's what the `map` is for below, to remove additional space around potential attributes:

```
attributes.map((attribute) => attribute.trim()).join(" ")
```

dcrousso replied:
looking more closely at `InspectorDOMAgent::setAttributesAsText` i dont think we actually even need to do this because it'll get handled for us in the backend

as such, we probably would be better off either only separating after the first whitespace
```js
let newTagName = newText;
let attributes = "";
let whitespaceIndex = newTagName.search(/\s/);
if (whitespaceIndex !== -1) {
    newTagName = newTagName.slice(0, whitespaceIndex);
    attributes = newTagname.slice(whitespaceIndex + 1).trim();
}

...


const name = "";
node.setAttribute(name, attributes, attributesAddedCallback);
```

btw we dont need to `newText.trim()` because that happens right before this

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -1271,18 +1271,37 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
-            var node = WI.domManager.nodeForId(nodeId);
+            let node = WI.domManager.nodeForId(nodeId);
+
+            if (attributes.length) {
```
id make this into an early return
```js
if (!attributes.length) {
    nodeUpdatedCallback(node, {moveFocus: true});
    return;
}

let attributesAddedCallback = (error, nodeId) => {
   ...
};

...
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -1271,18 +1271,37 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
+            }
+        }
 
+        let nodeUpdatedCallback = (node, moveFocus) => {
```
let's make this into `(node, options = {}) => {` so that we don't need to have a `const` at the callsite

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -1271,18 +1271,37 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
+                }
+
+                const attributeNameToReplace = "";
+                node.setAttribute(attributeNameToReplace, attributes.map(attr => attr.trim()).join(" "), attributesAddedCallback);
```
i wonder if we should update `WI.DOMNode.prototype.setAttribute` to have the `name` be part of an `options = {}` instead

Style: missing `(` `)`
Style: please don't abbreviate `attribute`

rcaliman-apple replied:
> i wonder if we should update `WI.DOMNode.prototype.setAttribute` to have the `name` be part of an `options = {}` instead

I'm inclined to leave it as is. It follows the expected arguments order of the DOM `setAttribute()` method and it looks right next to the getter. It's only a Web Inspector implementation detail that we can omit the attribute name in favor of letting the engine create new ones from the value string.

I'd much rather add a new `WI.DOMNode.prototype.setAttributesAsText()` that omits the `name` entirely for these use cases.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -1056,8 +1056,14 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
-            if (closingTagElement)
-                closingTagElement.textContent = "</" + tagNameElement.textContent + ">";
+            if (closingTagElement) {
+                let textContent = tagNameElement.textContent.trim();
```
the reason for `trimStart` is that we dont need to care about the end of the string because we'll either a) not find any (i.e. `whitespaceIndex === -1`) or b) slice to the last character before the whitespace, so it avoid unnecessary work

rcaliman-apple replied:
You're right! 

I made a mistake typing in your suggestion: `whitespaceIndex === -1` instead of `whitespaceIndex !== -1` 🤦

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -1271,30 +1277,50 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
+                    return;
+                }
+
+                nodeUpdatedCallback(node);
```
in the video it looks like if you hit enter with a bunch of attributes we just completely stop editing.  do we not want to focus the first attribute just in case the developer mistypes?  IMO that was kinda odd

rcaliman-apple replied:
The reasoning was that if you just typed in attributes, you don't need to go edit them straight away.

Focusing the first attribute is not guaranteed to focus what the author just typed in. Attributes are sorted alphabetically. We'd have to identify which was the first attribute in the string and switch that into edit mode. 
IMO, this is unnecessary. If there's a need to edit, the author can easily move into that mode.

dcrousso replied:
i suppose.  i kinda feel like we should at least focus _something_ as otherwise the developer has to take an extra step to re-select in order to navigate using tab (especially for things like screen readers), but i think we have similar behavior with some other interactions (e.g. contextmenu > Add > Attribute) so im fine with leaving this for now

rcaliman-apple replied:
> i kinda feel like we should at least focus something as otherwise the developer has to take an extra step to re-select in order to navigate using tab

That's a good argument. 👍  I lived on this for a bit and I agree that not moving focus makes tab navigation awkward. 

I added it back and changed it to select the first attribute if we added attributes with the tag name (instead of creating a new empty attribute when committing just a tag name).

Also added a drive-by fix to the automatic selection of attribute names when they're value-less (ex: `disabled`). Before, it caused an unhandled exception when tabbing through attributes because it tries to select the attribute value.

https://github.com/WebKit/WebKit/blob/76bd9ad78d4e916b082bd10c8346d287466216c7/Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js#L1018

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -1271,30 +1277,57 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
+        let whitespaceIndex = newText.search(/\s/);
+        if (whitespaceIndex !== -1) {
+            newTagName = newText.slice(0, whitespaceIndex);
+            attributes = newText.slice(whitespaceIndex + 1).trim();
```
NIT: i think we actually could just use `trimEnd()` because this will either be entirely whitespace (in which case `trim()` and `trimEnd()` will do the same thing) or it'll have content (and even if there is whitespace at the beginning it'll get handled for us in the backend).  not a huge deal tho as i am also a fan of trying to give a slightly nicer value to the protocol so we're sending less data across IPC/etc. :)

rcaliman-apple replied:
Turns out the backend isn't that resilient. If we pass it a string with extra space around each attribute it will return an `invalidCharacterError` from [InspectorDOMAgent::setAttributesAsText()](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/inspector/agents/InspectorDOMAgent.cpp#L783).

Likely from this iteration:

https://github.com/WebKit/WebKit/blob/26796239d990f7b51f4dc9418a764feb900d469b/Source/WebCore/inspector/agents/InspectorDOMAgent.cpp#L808-L814

(Waiting on a build with logs to confirm). 
The first attribute seems to be applied. See video:
 

https://github.com/user-attachments/assets/8bdb9246-b95f-479b-ad81-69c05317f6cb

  
I replaced `.trim()` with `.replace(/\s+/g, " ")` in the latest patch to address this.
 
 **Later edit:** Yes, the additional whitespace somehow ends up as a single-space attribute when doing:
https://github.com/WebKit/WebKit/blob/ae3be58163831c3aac676731a7eae002957b7b8a/Source/WebCore/inspector/agents/InspectorDOMAgent.cpp#L792

I'd rather not open that can of worms and just pass the sanitized attributes string from the frontend.

dcrousso reviewed (approved):
nice work!

rcaliman-apple committed:
Web Inspector: Parse attributes added via editing the tag name
https://bugs.webkit.org/show_bug.cgi?id=276533
rdar://problem/131607290

Reviewed by Devin Rousso.

Allow the user to input attributes when editing a node's tag name in the DOM node tree.

Take the first space-delimited substring to use as the tag name and use the rest as a string of attributes.
If the new tag name is invalid, the entire transaction will reject and revert to the original node.

* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype._startEditingTagName.keyupListener):
(WI.DOMTreeElement.prototype._tagNameEditingCommitted):
(WI.DOMTreeElement.prototype._tagNameEditingCommitted.changeTagNameCallback): Deleted.

Canonical link: https://commits.webkit.org/281507@main
